oauth2_http/java/com/google/auth/oauth2/InternalAwsSecurityCredentialsProvider.java

```diff
@@ -0,0 +1,282 @@
+    String token = (String) genericJson.get("Token");
+
+    // These credentials last for a few hours - we may consider caching these in the
+    // future.
+    return new AwsSecurityCredentials(accessKeyId, secretAccessKey, token);
+  }
+
+  @Override
+  String getRegion() throws IOException {
+    // If user has provided a region string, return that instead of checking environment or metadata
+    // server.
+    if (this.region != null) {
+      return this.region;
+    }
+    String region;
+    if (canRetrieveRegionFromEnvironment()) {
+      // For AWS Lambda, the region is retrieved through the AWS_REGION environment variable.
+      region = environmentProvider.getEnv(AWS_REGION);
+      if (region != null && region.trim().length() > 0) {
+        return region;
+      }
+      return environmentProvider.getEnv(AWS_DEFAULT_REGION);
+    }
+
+    Map<String, Object> metadataRequestHeaders = createMetadataRequestHeaders(awsCredentialSource);
+
+    if (awsCredentialSource.regionUrl == null || awsCredentialSource.regionUrl.isEmpty()) {
+      throw new IOException(
+          "Unable to determine the AWS region. The credential source does not contain the region URL.");
+    }
+
+    region = retrieveResource(awsCredentialSource.regionUrl, "region", metadataRequestHeaders);
+
+    // There is an extra appended character that must be removed. If `us-east-1b` is returned,
+    // we want `us-east-1`.
+    return region.substring(0, region.length() - 1);
+  }
+
+  private boolean canRetrieveRegionFromEnvironment() {
+    // The AWS region can be provided through AWS_REGION or AWS_DEFAULT_REGION. Only one is
+    // required.
+    List<String> keys = ImmutableList.of(AWS_REGION, AWS_DEFAULT_REGION);
+    for (String env : keys) {
+      String value = environmentProvider.getEnv(env);
+      if (value != null && value.trim().length() > 0) {
+        // Region available.
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  String getMetricsHeaderValue() {
+    return AWS_METRICS_HEADER_VALUE;
+  }
+
+  @Override
+  Supplier<AwsSecurityCredentials> getSupplier() {
+    return () -> {
+      try {
+        return this.getCredentials();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    };
+  }
+
+  private boolean canRetrieveSecurityCredentialsFromEnvironment() {
+    // Check if both AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are available.
+    List<String> keys = ImmutableList.of(AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY);
+    for (String env : keys) {
+      String value = environmentProvider.getEnv(env);
+      if (value == null || value.trim().length() == 0) {
+        // Return false if one of them are missing.
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @VisibleForTesting
+  boolean shouldUseMetadataServer() {
+    return (!canRetrieveRegionFromEnvironment()
+        || !canRetrieveSecurityCredentialsFromEnvironment());
```

Comment:
This can happen in the future, but:

now that you have an indirection to choose token providers, this logic can happen when installing a token provider and therefore break this class into one for environment and one for MDS.

Reply:
Unfortunately I think that would be a breaking change because the currently implementation checks the environment variables when get security credentials is called instead of in the constructor, so technically its possible for a user to create the credential and set environment variables after.